Feature/add transformation runs

fivetran-reneeli opened the pull request:
## PR Overview
**This PR will address the following Issue/Feature:** internal ticket

**This PR will result in the following new package version:** v1.11.0
<!--- Please add details around your decision for breaking vs non-breaking version upgrade. If this is a breaking change, were backwards-compatible options explored? -->

## Schema Changes: Adding the Transformation Runs Table
- We have added the `transformation_runs` source table. Note that not all customers have the `transformation_runs` source table, particularly if they are not using Fivetran Transformations. Therefore, the `transformation_runs` table will only populate if the table exists in your schema, via a new variable `fivetran_platform_using_transformations`, which automatically checks for the table. If the table doesn't exist, the staging `stg_fivetran_platform__transformation_runs` model will persist as an empty model and respective downstream fields will be null. 

- If the `transformation_runs` source table exists in your schema, `fivetran_platform_using_transformations` will be set to True and the following updates apply:
  - Added a new staging [`stg_fivetran_platform__transformation_runs`](https://fivetran.github.io/dbt_fivetran_log/#!/model/model.fivetran_log.stg_fivetran_platform__transformation_runs) model. 
    - We have also added the `get_transformation_runs_columns()` macro to ensure all required columns are present.
  - Added the following fields to the [`fivetran_platform__usage_mar_destination_history`](https://fivetran.github.io/dbt_fivetran_log/#!/model/model.fivetran_log.fivetran_platform__usage_mar_destination_history) end model for each destination and month:
    - `paid_model_runs`
    - `free_model_runs`
    - `total_model_runs`
  - If you would like to override these updates, you can also manually disable the `fivetran_platform_using_transformations` variable by setting it to False in your project.yml

## PR Checklist
### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [ ] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:
<!--- Provide the steps you took to validate your changes below. -->
hex notebook

### If you had to summarize this PR in an emoji, which would it be?
<!--- For a complete list of markdown compatible emojis check our this git repo (https://gist.github.com/rxaviers/7360908)  --> 
:dancer:

fivetran-jamie reviewed (changes requested):
This looks pretty good!! Made some minor suggestions and I have a couple of larger questions/requests:

1. Can we confirm with eng whether or not customers who do not use transformations at all will have the `transformation_runs` table in their schemas? Wondering if we need to include a variable
2. Can you move the Hex validations to a new `integrity` data validation test here?

fivetran-reneeli commented on models/staging/stg_fivetran_platform__transformation_runs.sql:
```diff
@@ -0,0 +1,42 @@
+                staging_columns=get_transformation_runs_columns()
+            )
+        }}
+        {{ fivetran_utils.source_relation(
+            union_schema_variable='fivetran_platform_union_schemas', 
+            union_database_variable='fivetran_platform_union_databases') 
+        }}
```
thanks for catching!

fivetran-reneeli commented on models/staging/stg_fivetran_platform__transformation_runs.sql:
```diff
@@ -0,0 +1,42 @@
+
+select
+    *,
+    {{ dbt.date_trunc('month', 'measured_date') }} as measured_month
```
Initially I had tried just having it just in the final CTE, but there were issues hence I did it here. Let me try it again though

fivetran-reneeli replied:
Oh yeah it's this error.
<img width="1024" alt="image" src="https://github.com/user-attachments/assets/185a7a70-cd49-464b-bb1b-f33a9e79e7cd" />

So it needs to go through the timestamp casting first otherwise it's not getting interpreted correctly.

fivetran-reneeli commented on models/fivetran_platform__usage_mar_destination_history.sql:
```diff
@@ -10,12 +10,24 @@ credits_used as (
+        cast(measured_month as date) as measured_month,
+        sum(case when free_type = 'PAID' then model_runs else 0 end) as paid_model_runs,
+        sum(case when free_type != 'PAID' then model_runs else 0 end) as free_model_runs,
+        sum(model_runs) as total_model_runs
```
added this in! thanks!

fivetran-reneeli commented on models/fivetran_platform__usage_mar_destination_history.sql:
```diff
@@ -10,12 +10,24 @@ credits_used as (
+
+    select
+        destination_id,
+        cast(measured_month as date) as measured_month,
```
moved it!

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,3 +1,18 @@
+## Schema Changes: Adding the transformation_runs table
+- We have added the `transformation_runs` source table. This includes the following updates:
+  - Added a new staging `stg_fivetran_platform__transformation_runs` model. Additionally, a new tmp model `stg_fivetran_platform__transformation_runs_tmp` and `get_transformation_runs_columns()` macro to ensure all required columns are present in the new staging model.
+  - Added the following fields to the `fivetran_platform__usage_mar_destination_history` end model for each destination and month:
```
added

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,3 +1,18 @@
+
+## Schema Changes: Adding the transformation_runs table
+- We have added the `transformation_runs` source table. This includes the following updates:
+  - Added a new staging `stg_fivetran_platform__transformation_runs` model. Additionally, a new tmp model `stg_fivetran_platform__transformation_runs_tmp` and `get_transformation_runs_columns()` macro to ensure all required columns are present in the new staging model.
```
added in and checked that the indenting looks good

fivetran-reneeli commented on models/staging/stg_fivetran_platform.yml:
```diff
@@ -187,4 +187,29 @@ models:
+        description: Boolean that indicates whether the user has verified their email address in the account creation process.
+
+  - name: stg_fivetran_platform__transformation_runs
+    description: Table of all transformation runs that have been executed by Fivetran.
```
thank you for catching!!

fivetran-reneeli commented on models/fivetran_platform__usage_mar_destination_history.sql:
```diff
@@ -63,6 +78,9 @@ join_usage_mar as (
         on destination_mar.measured_month = cast(usage.measured_month as date)
         and destination_mar.destination_id = usage.destination_id
+    left join transformation_runs
+        on destination_mar.measured_month = cast(transformation_runs.measured_month as date)
```
good point. i noticed both are previously already cast as date. let me try to remove them both

fivetran-reneeli commented:
Thanks @fivetran-jamie for the review, ready for another look! I also made the new validation test comparing run counts. I also asked eng, that's a good point.

fivetran-reneeli commented on models/staging/stg_fivetran_platform__transformation_runs.sql:
Now that we're using the `does_table_exist` macro and have to create a conditional block to pass all null fields for when table does not exist, I believe it doesn't work with the tmp model so I removed that, and instead wrapped up everything in the staging model as shown. This follows what's done in the `stg_fivetran_platform__usage_cost` model

fivetran-reneeli commented:
Updates are addressing the fact that this `transformation_runs` table doesn't always exist in the customer schema. Therefore I used the `does_table_exist` macro to then pass to the `fivetran_platform_using_transformations` variable. Major updates: removed the config from the src.yml, removed documentation about configuring the variable, changed the logic in stg_fivetran_platform__transformation_runs, and removed the conditional logic downstream now that the staging model will be passed regardless, just as an empty table if the source doesn't exist.

fivetran-jamie reviewed (commented):
runs great! some minor suggestions

fivetran-reneeli commented on models/staging/src_fivetran_platform.yml:
```diff
@@ -200,5 +200,28 @@ sources:
+
+      - name: transformation_runs
+        identifier: '{{ var("fivetran_log_transformation_runs_identifier", "transformation_runs") }}'
+        description: Table of all transformation runs that have been executed by Fivetran.
```
good note!

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,3 +1,23 @@
+    - `paid_model_runs`
+    - `free_model_runs`
+    - `total_model_runs`
+- *NOTE*: Not all customers will have the `transformation_runs` source table, particularly if they are not using Fivetran Transformations. Therefore, the `transformation_runs` table will only populate if the table exists in your schema. If the table doesn't exist, the staging `stg_fivetran_platform__transformation_runs` model will persist as an empty model and respective downstream fields will be ...
```
yeah true. i added it in here and shuffled the order of the sections

fivetran-reneeli commented on README.md:
```diff
@@ -85,12 +85,12 @@ vars:
 
 ### Step 4: Disable Models for Non Existent Sources
-If you do not leverage Fivetran RBAC, then you will not have the `user` or `destination_membership` sources. It's also possible you might not have any To disable the corresponding functionality in the package, you must add the following variable(s) to your root `dbt_project.yml` file. By default, all variables are assumed to be `true`:
+If you do not leverage Fivetran RBAC, then you will not have the `user` or `destination_membership` source tables. The `user` and `destination_membership` are enabled by default. Therefore in order to switch the default configurations, you must add the following variable(s) to your root `dbt_project.yml` file for the respective source tables you wish to disable:
```
added a section in the DECISIONLOG

fivetran-reneeli commented on models/fivetran_platform__usage_mar_destination_history.sql:
```diff
@@ -61,8 +76,12 @@ join_usage_mar as (
     from destination_mar 
     left join usage 
-        on destination_mar.measured_month = cast(usage.measured_month as date)
+        on destination_mar.measured_month = usage.measured_month
```
yes no prob!

fivetran-jamie commented on DECISIONLOG.md:
````diff
@@ -9,4 +9,7 @@ The below variables control the respective behaviors and may be overridden if de
+```
+
+## Transformation Runs
+Not all customers have the `transformation_runs` source table, particularly if they are not using Fivetran Transformations. Therefore, we leverage a new variable `fivetran_platform_using_transformations`, which automatically checks for the table.If it exists, the variable is set to True, which then persists the `transformation_runs` source table and related models and downstream fields. If the ta ...
````
```suggestion
Not all customers have the `transformation_runs` source table, particularly if they are not using Fivetran Transformations. Therefore, we leverage a new variable `fivetran_platform_using_transformations`, which automatically checks for the table. If it exists, the variable is set to True, which then persists the `transformation_runs` source table and related models and downstream fields. If the table doesn't exist, the staging `stg_fivetran_platform__transformation_runs` model will persist as an empty model and respective downstream fields will be null. 

In the case you have the `transformation_runs` source table but still wish to disable it to prevent it from being populated in the package, you may set `fivetran_platform_using_transformations` to False in your project.yml.
```

And can you add the variable yml configuration block, like what we have in the above section for `fivetran_platform__usage_pricing` and `fivetran_platform__credits_pricing`

fivetran-jamie commented on CHANGELOG.md:
```diff
@@ -1,3 +1,24 @@
+    - `free_model_runs`
+    - `total_model_runs`
+
+## Documentation Updates
```
can you add the DECISIONLOG addition to this section and link to it

fivetran-reneeli replied:
done!

fivetran-jamie commented on CHANGELOG.md:
```diff
@@ -1,3 +1,24 @@
+[PR #141](https://github.com/fivetran/dbt_fivetran_log/pull/141) includes the following updates:
+
+## Schema Changes: Adding the transformation_runs table
+- We have added the `transformation_runs` source table. Note that not all customers have the `transformation_runs` source table, particularly if they are not using Fivetran Transformations. Therefore, the `transformation_runs` table will only populate if the table exists in your schema, via a new variable `fivetran_platform_using_transformations`, which automatically checks for the table. If the  ...
```
Can you add a note somewhere in the CHANGELOG about being able to override this behavior by setting `fivetran_platform_using_transformations` to False, similar to what you have in the DECISIONLOG

fivetran-reneeli replied:
added

fivetran-jamie reviewed (commented):
looking great, couple of minor doc-related comments

fivetran-jamie reviewed (approved):
LGTM!

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,9 +1,28 @@
+
+- If the `transformation_runs` source table exists in your schema, `fivetran_platform_using_transformations` will be set to True and the following updates apply:
+  - Added a new staging [`stg_fivetran_platform__transformation_runs`](https://fivetran.github.io/dbt_fivetran_log/#!/model/model.fivetran_log.stg_fivetran_platform__transformation_runs) model. 
+    - We have also added the `get_transformation_runs_columns()` macro to ensure all required columns are present.
```
Let's move this to an under the hood change since this isn't really applicable for users as a breaking schema change.

fivetran-reneeli replied:
updated some of the language in the CHANGELOG

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,9 +1,28 @@
+
+## Documentation Updates
+- Included documentation about the `transformation_runs` source table and the aggregated `*_model_run` fields.
+- Added information about manually configuring the `fivetran_platform_using_transformations` variable in the [DECISION LOG.](https://github.com/fivetran/dbt_fivetran_log/blob/main/DECISIONLOG.md)## Documentation
```
```suggestion
- Added information about manually configuring the `fivetran_platform_using_transformations` variable in the [DECISION LOG](https://github.com/fivetran/dbt_fivetran_log/blob/main/DECISIONLOG.md).
```

fivetran-reneeli replied:
thanks for catching! removed

fivetran-joemarkiewicz commented on models/fivetran_platform__usage_mar_destination_history.sql:
```diff
@@ -10,12 +10,24 @@ credits_used as (
+    select
+        destination_id,
+        measured_month,
+        sum(case when free_type = 'PAID' then model_runs else 0 end) as paid_model_runs,
+        sum(case when free_type != 'PAID' then model_runs else 0 end) as free_model_runs,
```
We've seen cases in the past where the casing of strings in the Fivetran Platform data model could possibly change in the future. Can we make this more future proof and simple to a lower casing search?
```suggestion
        sum(case when lower(free_type) = 'paid' then model_runs else 0 end) as paid_model_runs,
        sum(case when lower(free_type) != 'paid' then model_runs else 0 end) as free_model_runs,
```

fivetran-jamie replied:
Hey I saw this was added but don't we do `upper(free_type) as free_type` in the staging model?

fivetran-reneeli replied:
Oh you are 100% correct thanks for noticing that. I think it makes more sense to keep casing in staging, but I just realized everywhere else we do a lower albet in the transforms. I'll do a `lower` in the staging and remove it from the transforms. 

cc @fivetran-joemarkiewicz

fivetran-joemarkiewicz commented on models/staging/src_fivetran_platform.yml:
```diff
@@ -200,5 +200,29 @@ sources:
+            description: Timestamp of when the record was last synced.
+
+      - name: transformation_runs
+        identifier: '{{ var("fivetran_log_transformation_runs_identifier", "transformation_runs") }}'
```
Small note - all other identifier variables have double quotes enclosing the bracket config and then single quotes for calling the variable. Let's use the same format as the other identifiers in this package to ensure consistency and that dbt doesn't behave any differently than expected across the board.
```suggestion
        identifier: "{{ var('fivetran_log_transformation_runs_identifier', 'transformation_runs') }}"
```

fivetran-joemarkiewicz commented on models/staging/stg_fivetran_platform__transformation_runs.sql:
```diff
@@ -0,0 +1,52 @@
+    cast(null as {{ dbt.type_timestamp() }}) as updated_at,
+    cast(null as date) as measured_month
+
+    {% if target.type not in ('sqlserver') %}
+    limit 0
+    {% endif %}
```
Do we need to account for Redshift to limit 1 as we've recently uncovered with the union data practices?

cc: @fivetran-jamie

fivetran-jamie replied:
Ah that's a great point! I think we should, as we've learned that Redshift won't respect the data type casts otherwise. @fivetran-reneeli can you add a condition to `limit 1` instead of 0 for Redshift targets

fivetran-reneeli replied:
Yes! I believe I'll need to add it to [usage_cost ](https://github.com/fivetran/dbt_fivetran_log/blob/4ebb0e6d4c1af92960e06c4283323df38d362c85/models/staging/stg_fivetran_platform__usage_cost.sql#L33) and [credits_used](https://github.com/fivetran/dbt_fivetran_log/blob/4ebb0e6d4c1af92960e06c4283323df38d362c85/models/staging/stg_fivetran_platform__credits_used.sql#L33) as well though. I'll add that in and note it in the CHANGELOG.

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,9 +1,28 @@
+    - `paid_model_runs`
+    - `free_model_runs`
+    - `total_model_runs`
+  - If you would like to override these updates, you can also manually disable the `fivetran_platform_using_transformations` variable by setting it to False in your project.yml
```
This CHANGELOG entry is confusing or misleading. I read this and imagine that if I define `fivetran_platform_using_transformations: false` then these three mentioned fields will not be present in the end model. However, setting this to false will simply ensure all of the results are `null`.

Can you confirm the `null` behavior is what is expected? If so, we should rethink this. Why would someone want to manually set this variable as `false` if the fields will still be present, just empty?

fivetran-reneeli replied:
Right it will be null if disabled. You're saying that it would be more likely that a customer disabling the variable would expect the fields not to be present at all, rather than null? The reason I did it this way was to mirror what is done for usage_cost, where if that table doesn't exist or is disabled, its downstream fields are all null. 

`fivetran_platform__usage_pricing: False`

![image](https://github.com/user-attachments/assets/2c9732e7-c67c-42d7-81c7-8a0ce89e637c)

fivetran-reneeli replied:
Updated the language to make it clearer that the fields will still be present but null.

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,9 +1,28 @@
 
-## Documentation
+## Schema Changes: Adding the Transformation Runs Table
+- We have added the `transformation_runs` source table. Note that not all customers have the `transformation_runs` source table, particularly if they are not using Fivetran Transformations. Therefore, the `transformation_runs` table will only populate if the table exists in your schema, via a new variable `fivetran_platform_using_transformations`, which automatically checks for the table. If the  ...
+
+- If the `transformation_runs` source table exists in your schema, `fivetran_platform_using_transformations` will be set to True and the following updates apply:
+  - Added a new staging [`stg_fivetran_platform__transformation_runs`](https://fivetran.github.io/dbt_fivetran_log/#!/model/model.fivetran_log.stg_fivetran_platform__transformation_runs) model. 
```
I feel we're adding more here than is necessary or useful to the customer and could potentially cause confusion to the end customer.

[Here](https://github.com/fivetran/dbt_fivetran_log/releases/tag/v0.6.0) is what we said when we introduced the usage_cost model.
![image](https://github.com/user-attachments/assets/58588f47-23e6-43e8-a2fc-7a1ad68106b9)

Can we update this entry to be a bit more concise and straight forward.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli a few questions and change requests before approval.

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,9 +1,27 @@
+## Schema Changes: Adding the Transformation Runs Table
+- This package now accounts for the `transformation_runs` source table. Therefore, a new staging model [`stg_fivetran_platform__transformation_runs`](https://fivetran.github.io/dbt_fivetran_log/#!/model/model.fivetran_log.stg_fivetran_platform__transformation_runs) has been added. Note that not all customers have the `transformation_runs` source table, particularly if they are not using Fivetran  ...
+
+- In addition, the following fields to the [`fivetran_platform__usage_mar_destination_history`](https://fivetran.github.io/dbt_fivetran_log/#!/model/model.fivetran_log.fivetran_platform__usage_mar_destination_history) end model:
```
```suggestion
- In addition, the following fields have been added to the [`fivetran_platform__usage_mar_destination_history`](https://fivetran.github.io/dbt_fivetran_log/#!/model/model.fivetran_log.fivetran_platform__usage_mar_destination_history) end model:
```

fivetran-joemarkiewicz commented on models/staging/src_fivetran_platform.yml:
```diff
@@ -200,5 +200,29 @@ sources:
+            description: Timestamp of when the record was last synced.
+
+      - name: transformation_runs
+        identifier: "{{ var('fivetran_log_transformation_runs_identifier', 'transformation_runs') }}"
```
Looks like we are using the old name here. We need to update this to platform.
```suggestion
        identifier: "{{ var('fivetran_platform_transformation_runs_identifier', 'transformation_runs') }}"
```

fivetran-joemarkiewicz replied:
We reference the platform name in the platform named version [here](https://github.com/fivetran/dbt_fivetran_log/blob/9f095e8939740dfe8cab0e6084aef5cc1736cbbb/integration_tests/dbt_project.yml#L23) as well, so this will be consistent

fivetran-joemarkiewicz reviewed (approved):
Two updates before release. Once those are updated this will be good to go!
